### DISCUSSION TITLE
Code review feedback

### DISCUSSION THREAD

**Discussion by ZZeller03** [frc5567/2023-Robot#151] [Team Posts]
~RobotMap.java
+Clarify that sensors are being plugged into DIO ports.

~StateMachine.java
+Fix all of the TODO: 's.  The comments for them aren't on the same line as the TODO:'s, so aren't being registered properly.

~DashBoardData.java
+Remove unused imports.
+Document the classes are in lines 24-28.
+Comment lines 50-54.

~ArduinoCommsStorage.java
+Clarify everything's purpose with more documentations.

~ArduinoToRioComms.java
+Remove unused imports.
+Document the classes.
+Command "==2" should be changed to a constant (line 103).
+Document lines 108-120.

~ClimbingMechanism.java
+Needs more documentation / comments for pretty much everything. 

~Controller.java.
+See Chris's example re-work code.

~CustomDashBoard.java
+Should the Enums be made into constants?
+Use Enum names instead of numbers [Ex. Hello=(1) Use Hello to refer to it and Not 1.  For clarification].
+Add methods to interface with DifferentialDrive rather than returning the object [Line 81].
+Should quick-rotate have a MAX_SPEED_DELTA?  (like with the acceleration speed).

~Elevator.java
+Should the Enums be in the be in the RobotMap?
+remove unused imports (line 7).
+The line:
```
		currentState = currentState.LEVEL_ZERO;
```
should be modified to 
```
		currentState = State.LEVEL_ZERO;
```

~GamePad.java
+Should GamePadControlls be in RobotMap?
+additional code Documentation needed (Specifically for GetX and GetY -- why do they need to be overridden like this?).

~HatchMech.java
+Should we use constants for the ports?
+Make comments into TODO: 's .
+Make speeds into constants.

~Robot.java
+The imports on line 36 may not be needed.
+Change the getLeftStickY & getRightStickX into getY(GenericHID.Hand.KLeft) & getX(GenericHID.Hand.KRight). and add import edu.wpi.first.wpilibj.GenericHID;

**Comment by JooshO** [comment]
In DuinoToRioComms.java
+There are no unused imports
+Line 103 is empty. The commands can be made into constants though

EDIT: Line 103 is lines 155- ~167. I'll update that
EDIT 2: The line 103 issue was actually in the Arduino code, but it should be fixed on my branch now
EDIT 3: Are any of these comments directed to the Arduino code or are all of them supposed to be for the .java

**Comment by coverbeek** [comment]
This is the feedback from the code review we did. In some cases, the feedback are considerations, in other cases we found issues and are reporting them here. Some of the line numbers already don't line up, unfortunately.

**Comment by coverbeek** [comment]
Here is the modified Controller.java that we propose is a slightly more elegant implementation.
The Controller class does not need to instantiate an XBoxController object, since it *IS* an XBoxController object. In the code below we removed that and directly override the XBoxController methods rather than add completely different methods.
```java
package frc.robot;

import edu.wpi.first.wpilibj.XboxController;

public class Controller extends XboxController {
    // Sets constants used for deadbands, one for sticks and the other for triggers
    private static final double STICK_DEADBAND = 0.05;
    private static final double TRIGGER_DEADBAND = 0.05;

    /**
     * Handles for input from driver's Xbox Controllers, and abjusts analog input using deadbands
     * @param port The Driver Station USB the controller is plugged into
     */
    Controller(int port) {
        super(port);
    }

    /**
     * Gets the value of the stick X axis
     * @return The X value of the stick axis
     */
    public double getX(Hand hand) {
        double stickXValue;
        // Checks if stick is within deadband
        if ((super.getX(hand) < STICK_DEADBAND) && (super.getX(hand) > -STICK_DEADBAND)) {
            // If so, stick value is set to 0
            stickXValue = 0.00;
        }
        // If the stick is outside deadband, sets variable to stick value
        else {
            stickXValue = super.getX(hand);
        }
        // Returns the stick value
        return stickXValue;
    }

    /**
     * Gets the value of the stick Y axis
     * @return The Y value of the stick axis
     */
    public double getY(Hand hand) {
        double stickYValue;
        // Checks if stick value is within deadband
        if ((super.getY(hand) < STICK_DEADBAND) && (super.getY(hand) > -STICK_DEADBAND)) {
            // If so, stick value is set to 0
            stickYValue = 0.00;
        }
        // If the stick is outside deadband, sets variable to stick value
        else {
            stickYValue = super.getY(hand);
        }
        // Returns the stick value
        return stickYValue;
    }

    /**
     * Gets the value of the trigger
     * @return The value of the trigger axis
     */
    public double getTriggerAxis(Hand hand) {
        double triggerValue;
        // Checks if the trigger value is less than deadband
        if (super.getTriggerAxis(hand) < TRIGGER_DEADBAND) {
            // If so, sets the trigger value to 0
            triggerValue = 0.00;
        }
        // If the trigger is outside of deadband, sets the variable to trigger axis
        else {
            triggerValue = super.getTriggerAxis(hand);
        }
        // Returns the trigger value
        return triggerValue;
    }
}
```

**Comment by patrick-philbin** [comment]
Need clarification on a suggestion for DashboardData.java. When you say "Document the classes are in lines 24-28",
what do you mean by this? Some line numbers have been changed so in my version these lines are all @param statements. Thanks!
@coverbeek

**Comment by coverbeek** [comment]
Something got lost in translation for that feedback.
For DashboardData.java I think we need documentation of the overall class and it's purpose. I think the comment on the constructor kind of explains it, but it would be helpful to have the documentation on the class itself.
The other comment was about putting some kind of explanation of what the proximity sensor formatting is doing and why.
